### PR TITLE
feat(sourceNamespace): Regex Support (#19016)

### PR DESCRIPTION
**Pull request by ArthurVardevanyan**
Here is a PR with my attempt at solving the below scenario: https://github.com/argoproj/argo-cd/issues/19016

It seems to work, however I may not have accounted for everything..

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Review by jannfis — COMMENTED** [review]
Hey @ArthurVardevanyan, thank you for trying to tackle this!

Instead of introducing a new command line option, WDYT of having a negation pattern (i.e. `!`) in the namespace list?

So, to specify all but a few namespaces, one could use:

`--application-namespaces *,!foo,!bar`

**Comment by ArthurVardevanyan** [comment]
> Hey @ArthurVardevanyan, thank you for trying to tackle this!
> 
> Instead of introducing a new command line option, WDYT of having a negation pattern (i.e. `!`) in the namespace list?
> 
> So, to specify all but a few namespaces, one could use:
> 
> `--application-namespaces *,!foo,!bar`

I agree, looks like the underlying glob library has a bug, otherwise this would have worked from day 1.
https://github.com/gobwas/glob/issues/47

What your thought be on just swapping it out (At least for the sourceNamespace portion), to using a proper regex library?

**Comment by jannfis** [comment]
> What would your thought be on just swapping it out (At least for the sourceNamespace portion), to using a proper regex library?

While I think it might be a good idea, you already mentioned that the change will be a breaking. I think the added value of using regexp is not sufficient to justify a breaking change. It might even have some weird side effects for people unaware of that change taking place, when their globs are now handled as regexps and could potentially introduce security issues or production outages in those environments. We can definitely keep that in mind for the next major version, though. If you'd like, you can go ahead and create an enhancement issue which I'll pin to the 3.0 milestone then.

Just snooping at what the changes in this PR look like, I think we're on a pretty good way now. I'll give it a thorough review tomorrow.

**Comment by ArthurVardevanyan** [comment]
> > What would your thought be on just swapping it out (At least for the sourceNamespace portion), to using a proper regex library?
> 
> While I think it might be a good idea, you already mentioned that the change will be a breaking. I think the added value of using regexp is not sufficient to justify a breaking change. It might even have some weird side effects for people unaware of that change taking place, when their globs are now handled as regexps and could potentially introduce security issues or production outages in those environments. We can definitely keep that in mind for the next major version, though. If you'd like, you can go ahead and create an enhancement issue which I'll pin to the 3.0 milestone then.
> 
> Just snooping at what the changes in this PR look like, I think we're on a pretty good way now. I'll give it a thorough review tomorrow.

I think I am content what the PR as it sits, I may add a few more tests for more scenarios, since technically this is not a "pure" glob implementation, want to make sure strange edge cases are accounted for.

However, it is not done like how the suggestion was:
```--application-namespaces *,!foo,!bar```
Since how the logic was written originally, match if any are true, I didn't want to deviate to much from that.

So I wrote it as, anything with a negate out front, just inverts the output from the glob library for that line item.

**Review by jannfis — COMMENTED** [review]
This test case I added fails:

```go
{
	"allow anything prefixed test, but not test-test",
	"test-test",
	"argocd",
	[]string{"!test-test", "test-*"},
	false,
},
```

I was thinking that maybe we can have a function that takes a list of patterns, maybe with a similar signature to: `MatchList(patterns []string, text string, separators ...rune)`. 

It will then build two lists, one having the positive globs, the second having the negative globs. Then it will:

* Match in the negative list. If there's a match, deny the namespace.
* Match in the positive list. If there's a match, allow the namespace. If there's no match, deny the namespace.

This would have the benefit that patterns do not need to be ordered (e.g. negatives first, positives last) and be a little more predictable.

**Comment by jannfis** [comment]
The following is a simple, non-optimized and untested example to illustrate the idea: https://play.golang.com/p/9JxiCrJM8XC

**Comment by ArthurVardevanyan** [comment]
That is interesting, could try that as well,

But 
What about a third option of supporting, regular strings, glob, and regex,


IE, if string starts with /, compute as regex ?

**Comment by jannfis** [comment]
> That is interesting, could try that as well,
> 
> But What about a third option of supporting, regular strings, glob, and regex,
> 
> IE, if string starts with /, compute as regex ?

That is actually a brilliant idea IMO! This will allow for a non-breaking change too, since `/` is not a valid character for a namespace and shouldn't be in use in any existing (working) glob pattern.

**Comment by ArthurVardevanyan** [comment]
> > That is interesting, could try that as well,
> > But What about a third option of supporting, regular strings, glob, and regex,
> > IE, if string starts with /, compute as regex ?
> 
> That is actually a brilliant idea IMO! This will allow for a non-breaking change too, since `/` is not a valid character for a namespace and shouldn't be in use in any existing (working) glob pattern.

Ended up doing with front and back, since that seems more standard.
https://regex101.com/
image

**Comment by jannfis** [comment]
Thanks. Is there a particular reason you chose `github.com/dlclark/regexp2` over Go's built-in `regexp` here?

**Comment by ArthurVardevanyan** [comment]
> Thanks. Is there a particular reason you chose `github.com/dlclark/regexp2` over Go's built-in `regexp` here?

I was running into this issue, and could not seem to get it to work with the added test case:
https://stackoverflow.com/a/42551653

**Comment by ArthurVardevanyan** [comment]
Testing with operator: https://github.com/ArthurVardevanyan/HomeLab/commit/f1843ceabe2bed2511d11593db05a6900440622a?diff=split&w=0

**Comment by ArthurVardevanyan** [comment]
Updated Doc

**Comment by ArthurVardevanyan** [comment]
> Sorry for the delay here, @ArthurVardevanyan and thanks for your patience.
> 
> I think it would make most sense to modify `MatchStringInList` to replace the `exactMatch bool` parameter with something that specifies the type of the match (i.e. exact, glob or regexp) and then use the appropriate matcher function accordingly for matching the tokens in the list.
> 
> `MatchStringInList` is used in different parts of the code base, and I'd not like to have those places change their behavior all at once.

Took a stab at this, take a look please!

**Review by jannfis — COMMENTED** [review]
You're quick :) Thanks. I only have a final comment now.

File: util/glob/list.go [hunk 1]
```diff
@@ -1,10 +1,24 @@
...
 	for _, ll := range list {
-		if item == ll || (!exactMatch && Match(ll, item)) {
+		// If string is wrapped in "/", assume it is a regular expression.
+		if patternMatch == "regexp" && strings.HasPrefix(ll, "/") && strings.HasSuffix(ll, "/") && regex.Match(ll[1:len(ll)-1], item) {
+			return true
+		} else if (patternMatch == "regexp" || patternMatch == "glob") && Match(ll, item) {
+			return true
+		} else if patternMatch == "exact" && item == ll {
 			return true
 		}
```
[the remark this answers is not in the document]
Comment (ArthurVardevanyan) [reply util/glob/list.go]: Updated to use constants .

**Review by jannfis — APPROVED** [review]
LGTM!

Thank you and congratulations on your first contribution to Argo CD, @ArthurVardevanyan :tada:
